arroyo/processing/strategies/batching.py

```diff
@@ -262,3 +265,265 @@ def create_with_partitions(
             self.__max_batch_size,
             self.__max_batch_time,
         )
+
+
+@dataclass(frozen=True)
+class MessageBatch(Generic[TPayload]):
+    """
+    Represents a batch of messages with the range of offsets covered.
+    The range of offsets is represented as a mapping between partitions
+    and offset ranges.
+    The lowest offset in each offset range is inclusize. The highest is
+    esclusive.
+
+    This class intentionally does not implement __len__ or __iter__ as
+    generally, when using it, we want to deal with None batch and empty
+    batches differently. This way it is less likely to make mistakes
+    when checking for an empty batch.
+    """
+
+    messages: Sequence[Message[TPayload]]
+    offsets: Mapping[Partition, OffsetRange]
+
+    def __repr__(self) -> str:
+        return f"<{type(self).__name__}: {len(self.messages)} message{'s' if len(self.messages) != 1 else ''}>"
+
+    def last(self) -> Optional[Message[TPayload]]:
+        return self.messages[-1] if self.messages else None
+
+    def is_empty(self) -> bool:
+        return len(self.messages) == 0
+
+
+class OutOfOrderMessage(Exception):
+    pass
+
+
+class BatchBuilder(Generic[TPayload]):
+    """
+    Accumulates messages in a `MessageBatch` object.
+    It requires offsets to be in monotonic order in each partition.
+    """
+
+    def __init__(
+        self,
+        max_batch_time: float,
+        max_batch_size: int,
+    ) -> None:
+        self.__max_batch_time = max_batch_time
+        self.__max_batch_size = max_batch_size
+        self.__messages: MutableSequence[Message[TPayload]] = []
+        self.__offsets: MutableMapping[Partition, OffsetRange] = {}
+        self.__init_time = time.time()
+
+    def append(self, message: Message[TPayload]) -> None:
+        if message.partition in self.__offsets:
+            if self.__offsets[message.partition].hi >= message.next_offset:
+                raise OutOfOrderMessage(
+                    f"Received offset {message.next_offset}. Current watermark {self.__offsets[message.partition].hi}"
+                )
```

Comment:
I would not bother with this. I think it's safe to assume messages are received in order. None of the other strategies do checks like this and I think it's unnecessary for everyone to be doing this.

Reply:
Removed